### PR TITLE
Refactor MatchAllQueryBuilder, TermQueryBuilder, IdsQueryBuilder

### PR DESCRIPTION
**Pull request by cbuescher**
MatchAllQueryBuilder, TermQueryBuilder, IdsQueryBuilder now include the fromXContent, doXContent and toQuery methods suggested in #9901. Also implementing the Streamable interface and adding unit tests for the serialization and parsing.

NOTE: this PR is against the feature/query-parse-refactoring branch.

### CODE REVIEW

File: src/test/java/org/elasticsearch/index/query/TermQueryBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+                        bind(CircuitBreakerService.class).to(NoneCircuitBreakerService.class);
+                    }
+                }
+        ).createInjector();
+
+        IndexQueryParserService queryParserService = injector.getInstance(IndexQueryParserService.class);
+        context = new QueryParseContext(index, queryParserService);
+        testQuery = createTestQuery();
+        String contentString = createXContent(testQuery).string();
+        parser = XContentFactory.xContent(contentString).createParser(contentString);
```
Comment (jpountz) [inline src/test/java/org/elasticsearch/index/query/TermQueryBuilderTest.java]: Maybe it would be easier to extend ElasticsearchSingleNodeTest so that all these bindings are already done for you?

Comment (cbuescher) [inline src/test/java/org/elasticsearch/index/query/TermQueryBuilderTest.java]: Thanks, I thought it nicer to have unit tests here than having to do the whole node setup. I'll try this, but how expensive is the setup of the ElasticsearchSingleNodeTest?
If we write more query tests like this I, an alternative would be to have an own BaseTest class for this. Already suggested this to Lee.

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java [hunk 2]
```diff
@@ -83,9 +79,27 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
...
+    }
+
+    public Query toQuery(QueryParseContext parseContext) {
+        if (boost == 1.0f) {
+            return Queries.newMatchAllQuery();
+        }
+
+        //LUCENE 4 UPGRADE norms field is not supported anymore need to find another way or drop the functionality
+        //MatchAllDocsQuery query = new MatchAllDocsQuery(normsField);
+        MatchAllDocsQuery query = new MatchAllDocsQuery();
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: this comment is there from ages. Given that normsField doesn't get parsed for ages either, we should just drop any reference of it I guess. Wondering if we should do that upstream though or here. Just scared that we make too many changes at the same time and we don't track them. Thoughts?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java [hunk 3]
```diff
@@ -100,15 +114,42 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
...
+            return false;
         }
 
-        //LUCENE 4 UPGRADE norms field is not supported anymore need to find another way or drop the functionality
-        //MatchAllDocsQuery query = new MatchAllDocsQuery(normsField);
-        MatchAllDocsQuery query = new MatchAllDocsQuery();
-        query.setBoost(boost);
-        return query;
+        MatchAllQueryBuilder other = (MatchAllQueryBuilder) obj;
+        return Objects.equals(boost, other.boost);
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: `return boost == other.boost` ?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java [hunk 4]
```diff
@@ -100,15 +114,42 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
...
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeFloat(this.boost);
+    }
+    
+    @Override
+    public int hashCode() {
+        int hash = super.hashCode();
+        hash = maybeHashcode(hash, boost);
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: seems odd that we have to call a method that accepts `Object` passing in a `float` as argument. Can we simplify?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java [hunk 5]
```diff
@@ -83,9 +79,27 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
 
     @Override
     public Query parse(QueryParseContext parseContext) throws IOException, QueryParsingException {
+        MatchAllQueryBuilder query = new MatchAllQueryBuilder();
+        query.fromXContent(parseContext);
+        return query.toQuery(parseContext);
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: I guess we could generify this on the base class since the last two lines are going to be the same everywhere? and have a new abstract method `newQuery()` that creates a new instance of the query(Builder) ?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java [hunk 6]
```diff
@@ -83,9 +79,27 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
...
+        MatchAllDocsQuery query = new MatchAllDocsQuery();
+        query.setBoost(boost);
+        return query;
+    }
+
+    public void fromXContent(QueryParseContext parseContext) throws IOException, QueryParsingException {
         XContentParser parser = parseContext.parser();
 
-        float boost = 1.0f;
+        this.boost = 1.0f;
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: nitpick, you might want to add `this.` to the below reference of boost during parsing (not shown here unfortunately)?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java [hunk 7]
```diff
@@ -64,15 +56,19 @@ public MatchAllQueryBuilder boost(float boost) {
         return this;
     }
 
+    /**
+     * @return the boost for this query.
+     */
+    public float getBoost() {
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: If we want to go with pure getter, which is fine with me, we should probably change our setters too then for consistency? I think we have `public MatchAllQueryBuilder boost(float boost)` here. Not talking about the fluent thing, just the name of the method, either get and set or no get nor set prefix .

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+                        bind(CircuitBreakerService.class).to(NoneCircuitBreakerService.class);
+                    }
+                }
+        ).createInjector();
+
+        IndexQueryParserService queryParserService = injector.getInstance(IndexQueryParserService.class);
+        context = new QueryParseContext(index, queryParserService);
+
+        testQuery = createTestQuery();
+        String contentString = createXContent(testQuery).string();
```
Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: `testQuery.toString()` ? then we could drop the `createXContent` method too?

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+    @Test
+    public void testToXContent() throws IOException {
+        XContentBuilder content = createXContent(new MatchAllQueryBuilder().boost(1.5f));
+        assertEquals(content.string(), MATCH_ALL_BOOST_1_5);
+    }
+
+    @Test
+    public void testFromXContent() throws IOException {
+        context.reset(parser);
```
Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: can we move this call to the setup method?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 7 above)
Comment (cbuescher) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: I also thought about making this package private, only used in test so far. Changing setter is also fine with me but will break client code somewhere I guess?

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+        terminate(injector.getInstance(ThreadPool.class));
+    }
+
+    @Test
+    public void testToXContent() throws IOException {
+        XContentBuilder content = createXContent(new MatchAllQueryBuilder().boost(1.5f));
+        assertEquals(content.string(), MATCH_ALL_BOOST_1_5);
```
Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: I would love to find a way to test this without having to compare strings, the reason being that a single whitespace breaks things. We could do this by introducing json libraries that do that looking at json content, but I wouldn't go there. How about creating our `QueryBuilder` as we already do in this test, calling `toXContent` on it, and then parsing its result back via `fromXContent` and comparing that the obtained QueryBuilder is equal to the original one?
I know this approach would mix up toXContent and fromXContext which is not very unit test friendly, on the other hand we would make sure that they work well together. With more complex queries we could for instance randomize the options that we set on them and just verify that given a certain input the output is the same. Thoughts?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 5 above)
Comment (cbuescher) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: Yes, I thought so. 
In the alternative I posted today I only have 

```
@Override
        public Query parse(QueryParseContext parseContext) throws IOException, QueryParsingException {
            return fromXContent(parseContext).toQuery();
       }
```

https://github.com/elastic/elasticsearch/compare/master...cbuescher:proto/queryrefactoring-matchall-v2#diff-b5958c2f636285e4231fcbb0ef38ebf5R167

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    @Test
+    public void testFromXContent() throws IOException {
+        context.reset(parser);
+        MatchAllQueryBuilder newMatchAllQuery = injector.getInstance(MatchAllQueryBuilder.class);
```
Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: why do we need to use guice to retrieve the builder? can't we create a new instance manually?

Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: I think I get it ow but I find it odd, maybe will clear up with following refactorings?

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        newMatchAllQuery.fromXContent(context);
+        // compare these
+        assertThat(testQuery.getBoost(), is(newMatchAllQuery.getBoost()));
```
Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: shouldn't we compare the two queries instead of just one of their fields? `assertThat(testQuery, equalTo(newMatchAllQuery));`?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 4 above)
Comment (cbuescher) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: This is copy from Lees PR, happy to simplify this in any way. Isabel mentioned utility functions in some google lib we are already using for this, will ask her about it.

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java [hunk 13]
```diff
@@ -83,9 +79,27 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
 
     @Override
     public Query parse(QueryParseContext parseContext) throws IOException, QueryParsingException {
+        MatchAllQueryBuilder query = new MatchAllQueryBuilder();
+        query.fromXContent(parseContext);
+        return query.toQuery(parseContext);
+    }
+
+    public Query toQuery(QueryParseContext parseContext) {
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: do we need the `QueryParseContext` argument here? we may need it in some other query?

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    @Test
+    public void testToQuery() throws IOException {
+        context.reset(parser);
+        MatchAllQueryBuilder newMatchAllQuery = injector.getInstance(MatchAllQueryBuilder.class);
+        newMatchAllQuery.fromXContent(context);
```
Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: maybe we shouldn't create a query by parsing it here, but only crate a new query builder and test the toQuery method against it instead of relying on `fromXContent`?

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+        BytesStreamOutput output = new BytesStreamOutput();
+        testQuery.writeTo(output);
+
+        BytesStreamInput bytesStreamInput = new BytesStreamInput(output.bytes());
+        MatchAllQueryBuilder deserializedQuery = new MatchAllQueryBuilder();
+        deserializedQuery.readFrom(bytesStreamInput);
+
+        assertEquals(deserializedQuery, testQuery);
+        assertNotSame(deserializedQuery, testQuery);
+        assertThat(createXContent(testQuery).string(), is(createXContent(deserializedQuery).string()));
```
Comment (javanna) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: I would probably take out the string comparison, once we know that the queries are equal, I think that's enough, we shouldn't test other methods on them, it's out of the scope of this test.

File: src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java (hunk 10 above)
Comment (cbuescher) [inline src/test/java/org/elasticsearch/index/query/MatchAllQueryBuilderTest.java]: Agreed on the part to avoid comparing to fixed strings. I included this just to have a simple point of reference for the creation of the XContent. We can drop this later I guess. I think what you suggest is what is done in the testFromXContent().

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 5 above)
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: I have yet to get to that alternative, I am lagging behind :)

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 7 above)
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: I think getters don't hurt and should have already been there. At the moment stuff can't be set but not retrieved, which seems silly. If we want to keep bw comp then let's leave the setter alone and follow the same convention with getters too

File: src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java [hunk 16]
```diff
@@ -204,18 +240,18 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
...
 
-        if (value == null) {
-            throw new QueryParsingException(parseContext.index(), "No value specified for term query");
-        }
-
+    public Query toQuery(QueryParseContext parseContext) {
+        Preconditions.checkNotNull(fieldName, "no fieldName specified for term query");
+        Preconditions.checkNotNull(value, "no value specified for term query");
         Query query = null;
         MapperService.SmartNameFieldMappers smartNameFieldMappers = parseContext.smartFieldMappers(fieldName);
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/TermQueryBuilder.java]: now I see why `QueryParseContext` here too. we should probably split the QueryParseContext in two data structures, one needed for parsing and one needed for toQuery (e.g. mapping etc.)

File: src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java [hunk 17]
```diff
@@ -37,18 +40,18 @@
...
+import java.util.Objects;
 
 /**
  * A query that will return only documents matching specific ids (and a type).
  */
-public class IdsQueryBuilder extends BaseQueryBuilder implements QueryParser, BoostableQueryBuilder<IdsQueryBuilder> {
+public class IdsQueryBuilder extends BaseQueryBuilder implements QueryParser, BoostableQueryBuilder<IdsQueryBuilder>, Streamable {
 
-    private final List<String> types;
+    private Collection<String> types = new ArrayList<String>();
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java]: new ArrayList<String>() => new ArrayList<>()

File: src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java [hunk 18]
```diff
@@ -37,18 +40,18 @@
...
 
-    private List<String> values = new ArrayList<>();
+    private Collection<String> values = new ArrayList<String>();
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java]: same as above

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 2 above)
Comment (cbuescher) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: Going to drop this since it's not supported in lucene and there seems to be nothing on this option in the docs, nor is that field parsed anywhere.

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 3 above)
Comment (cbuescher) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: Did this to be consistent with Lees PR here https://github.com/elastic/elasticsearch/pull/10405/files#diff-08a3154601eec8465b50ac68b45c4c48R523. 
Im happy to do this either way.

**Comment by javanna** [comment]
I left a bunch of comments. I am pretty positive that we can trim down the code needed for testing, by just having a base test class with a few abstract methods that provide the query to test etc. we'll see how far we can take this. I am happy that we are going with unit testing rather than requiring to fire up a node.

My personal preference, I think a single PR per query would be easier to review, also then you don't have to adapt all of the other queries to the same comments :) but I see how this time it was better to touch a few of them rather than just one.

File: src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java [hunk 19]
```diff
@@ -88,6 +94,10 @@ public IdsQueryBuilder boost(float boost) {
         return this;
     }
 
+    public float getBoost() {
```
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java]: one more thing, these are going to be user facing public methods, we should add javadocs for each of them

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 2 above)
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: agreed on dropping, which we kinda did already :) how do proceed though? upstream? in our branch? how we do we track of these unrelated changes (I fear there's going to be many given that the deeper you dig the more you find...)?

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 3 above)
Comment (javanna) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: I see now... yea it's odd here cause this query has a single float field, looks better on more complex queries (especially cause you don't really see that among many fields)...

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 2 above)
Comment (cbuescher) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: I can open issue for this upstream, discuss and fix it there, then revert this change here. Will be merged then later.

File: src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java (hunk 3 above)
Comment (cbuescher) [inline src/main/java/org/elasticsearch/index/query/MatchAllQueryBuilder.java]: I'll leave this in for now, since it also takes care of null checks (although that shouldn't be possible for float boost here)

**Comment by cbuescher** [comment]
I worked through your comments. When we got through the other PR with Lees query we can start identifying common issues and test setup that can be generalised. I'll wait with further improvements here till we decided what kind of Parser/Builder merge we want on the feature branch and if we have to rebase there.

**Comment by javanna** [comment]
This got replaced by individual PRs, one per query: #10668 , #10669 & #10670 . Closing.
